pkg/controller/issueapi/handle_issue.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package issueapi
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/google/exposure-notifications-verification-server/pkg/api"
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/observability"
+	"github.com/google/exposure-notifications-verification-server/pkg/rbac"
+)
+
+// HandleIssueAPI responds to the /issue API for issuing verification codes
+func (c *Controller) HandleIssueAPI() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if c.config.IsMaintenanceMode() {
+			c.h.RenderJSON(w, http.StatusTooManyRequests,
+				api.Errorf("server is read-only for maintenance").WithCode(api.ErrMaintenanceMode))
+			return
+		}
+
+		startTime := time.Now()
+		if result := c.IssueWithAPIAuth(w, r); result != nil {
+			recordObservability(r.Context(), startTime, result)
+		}
+	})
+}
+
+// HandleIssueUI responds to the /issue API for issuing verification codes
+func (c *Controller) HandleIssueUI() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if c.config.IsMaintenanceMode() {
+			c.h.RenderJSON(w, http.StatusTooManyRequests,
+				api.Errorf("server is read-only for maintenance").WithCode(api.ErrMaintenanceMode))
+			return
+		}
+
+		startTime := time.Now()
+		if result := c.IssueWithUIAuth(w, r); result != nil {
+			recordObservability(r.Context(), startTime, result)
+		}
+	})
+}
+
+func (c *Controller) IssueWithAPIAuth(w http.ResponseWriter, r *http.Request) *IssueResult {
+	ctx := r.Context()
+	result := &IssueResult{
+		HTTPCode:  http.StatusOK,
+		obsResult: observability.ResultOK(),
+	}
+
+	if authorizedApp := controller.AuthorizedAppFromContext(ctx); authorizedApp == nil {
+		result.obsResult = observability.ResultError("MISSING_AUTHORIZED_APP")
+		controller.MissingAuthorizedApp(w, r, c.h)
+		return result
+	}
+
+	c.decodeAndIssue(ctx, w, r, result)
+	return result
+}
+
+func (c *Controller) IssueWithUIAuth(w http.ResponseWriter, r *http.Request) *IssueResult {
+	ctx := r.Context()
+	result := &IssueResult{
+		HTTPCode:  http.StatusOK,
+		obsResult: observability.ResultOK(),
+	}
+
+	if membership := controller.MembershipFromContext(ctx); membership == nil || !membership.Can(rbac.CodeIssue) {
```

Comment:
If the UI Auth was called, the middleware should have set the membreship, so a missing membership should be an internal error. But a lacking permission should still be unauthorized.